### ISSUE TITLE
[css-conditional] [css-contain] `srcset` and `sizes` interaction with container queries

### ISSUE THREAD

**Issue by una** [w3c/csswg-drafts#5889]
With container queries, the way developers write responsive styles changes from querying the global page for information to querying for a parent with containment for information. This shifts responsive paradigms. One case I can forsee is the need to query the parent for sizing and placement of its children. This includes `srcset` and `sizes` for [responsive images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images)

See also, the need for `cw` and `ch` units: #5888

Would the following be able to work with container queries instead of global queries? (ignore syntax, just trying to show the idea):

```
<img srcset="elva-fairy-480w.jpg 480w,
             elva-fairy-800w.jpg 800w"
     sizes="container(max-width: 600px) 480px,
            800px"
     src="elva-fairy-800w.jpg"
     alt="Elva dressed as a fairy">
```

```
<picture>
  <source container="(max-width: 799px)" srcset="elva-480w-close-portrait.jpg">
  <source container="(min-width: 800px)" srcset="elva-800w.jpg">
  <img src="elva-800w.jpg" alt="Chris standing up holding his daughter Elva">
</picture>
```

**Comment by chrishtr** [comment]
@yoavweiss @cramforce WDYT?

**Comment by chrishtr** [comment]
@mirisuzanne

**Comment by cramforce** [comment]
@chrishtr The use case I was mentioning earlier basically the opposite. Containers often determine the width of their child. So what I want is:

```html
<main sizes="50vw">
  <img srcset="1.avif 1920w, 2.avif640w"><!-- no redundant sizes attribute on children of main -->
  <img srcset="1.avif 1920w, 2.avif 640w">
  <img srcset="1.avif 1920w, 2.avif 640w">
</main>
```

**Comment by mirisuzanne** [comment]
I completely agree that this would be a very powerful and useful feature. My understanding was that responsive image sources are resolved before styles – making this reference impossible. But I'm not a browser engineer, and would love to discover that I'm wrong about this. :)

**Comment by scottjehl** [comment]
Can't speak to the implementation side, but I like this idea, @una. Just like with container queries, I think the container size is what we typically want to know for specifying image sources. Also, in the off chance that `media` [comes back to video](https://www.filamentgroup.com/lab/video-with-sizes/), it'd be nice to see this working there as well.

**Comment by eeeps** [comment]
This seems **really** exciting for art direction (e.g., `<picture><source container>` or whatever).

However, part of the goal of `<source media>` and `srcset` + `sizes` is to allow UAs to load images _immediately_, using information contained within markup alone, and I don't see how you resolve a `cw` or a `container()` or whatever, without waiting for style and layout.

For art direction... maybe that's a high performance price we have to pay to enable art direction within responsive components, which seems important.

But for `sizes` and `srcset` selection, once you've waited for layout, you don't need `sizes` at all. (See also: https://github.com/whatwg/html/issues/4654)

I can maybe see a path forward to instant loading here if the queried container, itself, has a `sizes` attribute, which can be resolved using just viewport size... e.g.

```
<div
  class="container"
  sizes="(min-width: 800px) 800px, 100vw" />
  <img
    srcset="small.jpg 400w, medium.jpg 800w, large.jpg 1200w"
    sizes="container(min-width: 600px) 33cw, 100cw"
  />
  [more images with the same sizes go here]
</div>
```

But writing `sizes` is already extremely difficult; this figures to make it worse.

**Comment by chrishtr** [comment]
> I don't see how you resolve a `cw` or a `container()` or whatever, without waiting for style and layout.

Under this proposal, the selection of the size would indeed depend on style and layout. Which would mean the loading of these images would be significantly delayed. The same problem occurs with `loading=lazy`, because the only way to determine if
the image intersects the viewport is to do layout.

**Comment by Wilto** [comment]
> `srcset` and `sizes` interaction with container queries

_One loves to see it_.

I think @joemcgill nailed my thoughts on this, but one thing worries me a little: “lazy load images based on container width” feels like both a sensible default for omitted `sizes` and a powerful feature in the right hands—like, “I want lazyload on this `img` in this CQ, so I’ll skip writing `sizes`” sounds great on a couple of levels.

What worries me is that `sizes` attributes (and please know I mean this with all love for the feature) are a huge pain to write by hand. Understanding the difference between preparse and layout calculation timings is a lot to ask of a junior developer that just wants to put an image on a page, and asking devs to calculate an accurate `sizes` attribute in order to shave off imperceptible-in-the-moment differences in load time could be a tough sell. “You can go to all the trouble of writing a `sizes` attribute and the image shows up quick and looks right, or you can not write `sizes` and the image still shows up quick and looks fine,” y’know?

I’m a little concerned that the lazyloading pattern would end up telephone-gamed into “we don’t need `sizes` in CQ,” and we’d be effectively giving up preload on `srcset`-powered images inside CQ. That said, `srcset` has always felt a little like a stopgap. So, hm.

**Comment by MattWilcox** [comment]
`That said, srcset has always felt a little like a stopgap`

Yep. When we have aspect-ratio on hand and available, plus loading=lazy, then we can safely let the browser decide which asset to pull because *it already knows the width it actually needs all by itself* without hinting.

**Comment by yoavweiss** [comment]
*belatedly catching up...*

I agree with @eeeps about the inherent tradeoff here, about the ability to go with nested `sizes` and about the inability of humans to make sense of that.

I similarly agree with @chrishtr that for anything outside any initial viewport, lazy loading it has no performance cost and should be encouraged.

But, that leaves us with "images inside a container query that are in an initial viewport of some form factors".
What would be our developer advice in that case?
* The initial layout is fast if the relevant CSS is fast to arrive. Can we simply advise developers to make sure the layout structure of their initial viewport is laid out by inline styles? Is that a reasonable thing to ask?
* While humans won't be able to generate nested `sizes` attributes, tools might. *&lt;insert "the tools will save us" meme here>* Is that a reasonable path?
* We tell developers to avoid Container Queries in the initial viewport? (even if we would, they are not likely to listen)
* We live with slower LCPs for developer convenience? (Not a fan of this option, but might be interesting to quantify the damage IRL)
* A better option that has alluded us for the last 6 years?

I'm not sure what the answer is, but seems like we need one.

**Comment by rik** [comment]
> The initial layout is fast if the relevant CSS is fast to arrive. Can we simply advise developers to make sure the layout structure of their initial viewport is laid out by inline styles? Is that a reasonable thing to ask?

I don't think it's reasonable. Maintaining an accurate set of inline styles for the initial viewport is a very hard task and I've never met a team who does it well. (Sidenote: it is very annoying to see this recommandation in performance tooling as it is quite unrealistic)

***

Currently, responsive images can be fetched before the external CSS has been parsed. Changing this behaviour because of the introduction of container queries seems like a breaking change to me. I think browsers could keep the existing behaviour if no condition in sizes targets a container. Using the draft syntax:

```html
<!-- Starts fetching with only HTML known -->
<img srcset="elva-fairy-480w.jpg 480w,
             elva-fairy-800w.jpg 800w"
     sizes="(max-width: 600px) 480px,
            800px"
     src="elva-fairy-800w.jpg"
     alt="Elva dressed as a fairy">
```

```html
<!-- Waits for external CSS to choose an image to fetch -->
<img srcset="elva-fairy-480w.jpg 480w,
             elva-fairy-800w.jpg 800w"
     sizes="container(max-width: 600px) 480px,
            800px"
     src="elva-fairy-800w.jpg"
     alt="Elva dressed as a fairy">
```

**Comment by chrishtr** [comment]
> * The initial layout is fast if the relevant CSS is fast to arrive.

I think this is reasonable to aim for a first layout that runs fast enough. Recent features such as `content-visibility: auto` make it feasible to make large documents have fast first layout.

This leaves the question of whether it is reasonable to expect the relevant CSS to be fast to arrive. This would require critical styles to be inlined into the HTML, or something like Early Hints, if "fast to arrive" means no extra round-trips.

**Comment by bkardell** [comment]
A few of us are planning a breakout to discuss this topic/use cases on Thursday March 11 at 9am ET, the invitation for anyone intersted in joining that is open.. Please reach out to @una or myself and we'll share a meeting link/calendar invite.

**Comment by una** [comment]
Informal breakout meeting notes: https://docs.google.com/document/d/1AgCqgnPymQPzwdl0-BgNd22Sc01xXebjVezkme2ofu4/edit?usp=sharing

**Comment by astearns** [comment]
@una can you export the google doc into an archival format and send it as an attachment to www-archive? Then we can have a more permanent link to add here.

**Comment by jakearchibald** [comment]
I have a hunch that many (most?) `sizes` in the wild are wrong, and users are getting oversized images as a result. Eg, a lot of the `sizes` in https://news.sky.com/uk are just 100vw.

Auto sizes might be a net win for performance, even if it delays image loading until after layout.

Call it `sizes="delayed-auto"` or whatever if you want devs to think twice about using it. But for stuff below the fold (which may already be using `loading="lazy"`, it doesn't seem like a big deal.

**Comment by MattWilcox** [comment]
Fwiw I don't think the use case of "request an asset as fast as possible" is anything like as important as "load the right asset and avoid wasting potentially hundreds of Kb of inappropriate image data". 

The part of the feature that allows for early http requests is far less impactful for any page of a realistic size than getting the more appropriate asset for the context in use. Given that the majority of the time "saved" by this behaviour is actually dns lookup time and/or connection establishment... those are mostly redundant? A huge portion of images are going to be on the same domain or a CDN that don't incur an extra lookup, or over http2/3 which don't suffer the same connection overhead.

Unless I'm missing something? I'd happily ditch the "load a thing as soon as possible" for a "load the right thing".

**Comment by jakearchibald** [comment]
> Given that the majority of the time "saved" by this behaviour is actually dns lookup time and/or connection establishment...

Ohh, on that note, if all the potential sources for an `<img>` are the same origin, but not the same origin as the page, the browser can start that connection before layout.

I think this might be a common case. Eg images on https://www.theguardian.com/uk are on a different origin to the page, but the images themselves are all on the same origin as each other.

**Comment by jakearchibald** [comment]
Ugh, unless the image turns out to be `display:none` or within `display:none`

**Comment by MattWilcox** [comment]
> Ugh, unless the image turns out to be `display:none` or within `display:none`

True, though I'd think that's not any different than how that scenario would be today anyway?

**Comment by eeeps** [comment]
>  I'd happily ditch the "load a thing as soon as possible" for a "load the right thing".

One of the core design constraints for `srcset` (almost a decade ago!?) was that it not add any delay to load start. The strongest argument for that IMO was "a performance feature shouldn't slow down your image loads".

Tying layout-aware responsive image loading to lazy-loading makes sense, as we're not adding any additional delay. Tying it to container queries makes less sense, from this point of view. But it would be nice to quantify both:

1. The cost of the author-facing complexity required to allow pre-layout load starts (here I'm mostly talking about `sizes`); how many authors are doing responsive image markup badly and what impact is that having on users?
2. The potential user-facing performance costs of waiting for layout to select responsive image resources.

I'm doing [some analysis on #1 as part of the HTTP Almanac this year](https://github.com/HTTPArchive/legacy.httparchive.org/pull/220), and am very excited about it. I have some A/B demos that investigate #2 synthetically, but don't have good ideas about how to extend that analysis to the web, as it exists, as a whole.

**Comment by zcorpan** [comment]
Also see: https://github.com/whatwg/html/issues/8007

**Comment by grigs** [comment]
I’m a bit late to this discussion, and I’m not certain how much of a contribution this will be to finding a solution, but I wrote [an article](https://cloudfour.com/thinks/on-container-queries-responsive-images-and-jpeg-xl/) about the challenge and my feeling that maybe this is the time when we should revisit some of the assumptions that led us to the responsive images syntax.

**Comment by rjgotten** [comment]
Has any of the discussion for this considered looking at the problem from a different angle?

`srcset` had one major design flaw, or rather, two related ones:
1. no constraint on the images in the set being required to be same-aspect; and
2. no a-priori knowledge of the aspect ratio.

Introduce that constraint and introduce an attribute that can provide that aspect ratio and suddenly for initial viewport rendering you can speculatively download whatever is smallest. If once layout happens the image hasn't been downloaded yet, then you can abort the pending request and transparently upgrade to the appropriate sized image.
If once layout happens and the image _has_ been downloaded, start the upgrade request anyway, but at least you'll have a crude placeholder to use for first paint. A placeholder that can be transparantly replaced later, because of the constraint on the images being same-aspect - i.e. you *know it will not shift layout.*
 
Doesn't even have to be the smallest defined image in the srcset.
Could be a separate property altogether. Like how `<video>` has a `poster` attribute.

**Comment by zcorpan** [comment]
@rjgotten I'm not sure it's acceptable to load lowest res image first and after layout load the appropriate image. I think that would cause "double download" of all images in the general case, which is generally seen as a bug. Ideally only the appropriate image is loaded, without blocking on layout.

@grigs For new image formats, as far as I can tell the art direction use case is not handled. With the bookstore demo also it's a completely different image, not just a crop. I suppose an image file could in theory contain different images though.

**Comment by pc-erin** [comment]
@zcorpan Isn't the `picture` element used for art direction and format handling? You could do art direction with srcset and sizes, but from what I understand it wasn't intended for that.

**Comment by zcorpan** [comment]
@pc-erin yes, but it's limited to media queries.

**Comment by grigs** [comment]
> @grigs For new image formats, as far as I can tell the art direction use case is not handled. With the bookstore demo also it's a completely different image, not just a crop. I suppose an image file could in theory contain different images though.

Not that it matters for your greater point, but the bookstore demo is the same image with CSS effects applied to it in different contexts. 

`<picture>` likely requires some form of mapping the queries on the sources inside the container to the size of the container and the something on the container itself to the viewport. 

So the art direction use case would still require `<picture>` and `<source>` options, but progressive downloading of an image could remove the need for `sizes` and multiple options in the `srcset`.

It seems that developers have had the most trouble using `sizes` and `srcset` correctly. And unlike the problems with developers setting `sizes` values incorrectly, when I see `<picture>` misused, it is usually scenarios where `sizes` and `srcset` should be used instead. 

Web Almanac says [`<picture>` is used 7.7% of the time](https://almanac.httparchive.org/en/2022/media#picture). So if we can simplify the resolution switching use case, we’re going to simplify over 90% of responsive images.

**Comment by eeeps** [comment]
@grigs:
> Web Almanac says [<picture> is used 7.7% of the time](https://almanac.httparchive.org/en/2022/media#picture). So if we can simplify the resolution switching use case, we’re going to simplify over 90% of responsive images.

- 7.7% of pages used `<picture>`
- 34% of pages used `<img srcset>`

So it's closer to 80/20. Although I think what we really want to compare is `w` descriptor vs `media` attribute usage, and that isn't really derivable from the published data, but 80% "flexible image use case" / 20% "art direction use case" is a decent first approximation.

**Comment by austinh** [comment]
Engineer at Airbnb here. We push the limits on quality and bandwidth for images and we find media queries not specific enough for more complex layouts (grids that flow based on container, masonry rtc) and it would be great to have a declarative syntax that lets us get specific about the container so we can "right size" an image, because ultimately, thats what an image often fills - a container, rarely a screen. Media queries are blunt tools and we need a scalpel not a hammer. I wonder if there is some world where we can expand upon the picture source sets for container queries that strikes a balance between performance (preload scanner) and right sizing images (for quality, bandwidth, and art direction).

We attempted to solve this at the JS or service worker layer and found it needs to be solved declaratively at the browser level.

Let me know if there are any questions about our use cases.

**Comment by emilio** [comment]
I'd be curious @austinh, does [`sizes=auto`](https://html.spec.whatwg.org/#valdef-sizes-auto) fit your use-case? If not, why not?

**Comment by austinh** [comment]
> I'd be curious @austinh, does [`sizes=auto`](https://html.spec.whatwg.org/#valdef-sizes-auto) fit your use-case? If not, why not?

It does not for a few reasons:

* art direction : we may want to show a cropped square on smaller screens for example
* hero images - it requires every image to be lazy loaded.
* afaik img sizes="auto" needs to have the width and height up front but we are almost always working with images that are scaled to fill the container and dont know this.

**Comment by emilio** [comment]
> afaik img sizes="auto" needs to have the width and height up front but we are almost always working with images that are scaled to fill the container and dont know this.

you need `width` and `height` of the _original_ image, not of the scaled one, right?

**Comment by austinh** [comment]
> > afaik img sizes="auto" needs to have the width and height up front but we are almost always working with images that are scaled to fill the container and dont know this.
> 
> you need `width` and `height` of the _original_ image, not of the scaled one, right?

Ah yes, you're right. In that case that point is mute, but the other two are still whats preventing us from using lazy to solve this.

The other additional DX issue is that many devs are using container queries now in css for the layout and have a hard time describing their img/picture media queries that match up to their css. 

Is there not some way to have a loading="layout" that is after preload scanner but before lazy loading? Or perhaps have preload scanner pre calculate a "best guess" at the layout within a reasonable time frame?

**Comment by emilio** [comment]
@austinh I think at least the second one is kinda unsolvable? Or what am I missing? You do need to do layout to determine which image to load, so by definition you can't load the image eagerly, right? You could prefetch or something, I guess, but...

**Comment by emilio** [comment]
Regarding the art direction use case... Does the srcset not resolve against the container size? If so, wouldn't that allow you to address it? (Might be misremembering the interaction, but I thought that use case was covered in the design, cc @zcorpan )

**Comment by austinh** [comment]
> @austinh I think at least the second one is kinda unsolvable? Or what am I missing? You do need to do layout to determine which image to load, so by definition you can't load the image eagerly, right? You could prefetch or something, I guess, but...

I guess Im asking if there is any suggested solution to "right size" images for hero or even just initial viewport images - but say you don't know the size upfront because it's expressed in container query. Sizes=auto always requires a lazy load.

Is lazy really the soonest the browser can load this information? I saw a proposal at some point for after-layout - something in between eager and lazy. This would allow for better quality images but also better performance (not over fetching for example).

**Comment by JoshTumath** [comment]
> Is lazy really the soonest the browser can load this information? I saw a proposal at some point for after-layout - something in between eager and lazy. This would allow for better quality images but also better performance (not over fetching for example).

It's the soonest because the HTML parser loads an image as soon as it identifies an image source URL - it doesn't wait until a later phase, so - like Emilio said - it's downloaded before the browser has even started to process any CSS, so we can't wait until after layout unless you use `loading=lazy`.

Do you mean there was a proposal to add an option to the loading attribute like `loading=eager-after-layout`? That would certainly also allow for what you want.

We have a similar problem in #13377.

**Comment by eeeps** [comment]
@austinh @JoshTumath I think https://github.com/whatwg/html/issues/8007 is the relevant issue for `loading=eager-after-layout` (or whatever).
